### ISSUE TITLE
Design thoughts

### ISSUE THREAD

**Issue by mlutze** [mlutze/flixball#40]
Some assorted ideas about the code structure, etc:
1. We should just use the `Coordinates` type alias everywhere instead of separate rows and cols fields. Only very rarely do we actually need to unpack it.
2. `Display` should not know anything about the AI. I think instead of passing it a game state or a board, we should construct a `DisplayBoard` that contains the minimum information to display the board. This way the APIs stay separate.
3. It would be cool if AIs were combinable/composable. For example, I have an composite AI that tries AI1 35% of the time or else it tries AI2. Or if AI1 doesn't know what to do, we fallback to AI2. Maybe that's already possible.
4. The ugliness juggling states in `Engine` might be cleaned up with either a monad or internal mutability. Probably not internal mutability though because then we'd need a mutable GameState, I think?

@JonathanStarup comments?

**Comment by JonathanStarup** [comment]
1. yes, my code has been missing some reorganization. I think all of our data structuring could use a rethink to avoid hassle, coupling, and redundancy.
2.

**Comment by mlutze** [comment]
Here's my current thought on GameState etc.

GameState currently has
- `states: Map[PlayerId, (Int32, Map[String, String])]`
- `positions: Map[PlayerId, Position]`
- `board: Board`

And `Board` contains `Tile`s which contain `Player`s which contain `MoveLogic`

I think we should essentially get rid of `Player` and instead store the relevant data in the maps.

So `GameState` becomes `(states, positions, logics, board)`. And things dealing with `Player` instead deal with `PlayerId`.

Then we separate GameState into `PublicState` and `PrivateState`. This way we keep the states and logic of all the AIs separate from each other. So the `getMove` function for AIs has the signature: `(Int32, Map[String, String]) -> PublicState -> ((Int32, Map[String, String]), Move)`

**Comment by JonathanStarup** [comment]
I agree generally.
* I still think the MoveLogic should be regioned, which removes the Int32, Map... and States map which simplifies things.
* Our running of the board has become stateful but with the above change, why is it stateful? the positions map is just there for performance right? it can be computed from the board

**Comment by mlutze** [comment]
>     * I still think the MoveLogic should be regioned, which removes the Int32, Map... and States map which simplifies things.

Yes, I agree if it can be made to work, but I am not yet convinced it can work :)

>     * Our running of the board has become stateful but with the above change, why is it stateful? the positions map is just there for performance right? it can be computed from the board

Positions is indeed redundant. I don't totally understand what you mean by stateful though. Here I'm mostly just rearranging where the data is located.

**Comment by mlutze** [comment]
I've run into a problem with the Impure to Pure MoveLogic function.
```scala
    type alias MutMoveLogic[m: Region -> Type, r: Region] =
        (GameState, PlayerId, m[r]) -> Move \ r

    def toMoveLogic(logic: MutMoveLogic[m, r]): MoveLogic with State[a] = {
        def f(gs, id, state) = region reg {
            // call logic() with conversions
        }
    }
```

The issue is that here, we want the `r` of `MutMoveLogic[m, r]` to unify with `reg`. But `r` is already bound by the function. For example we could call `toMoveLogic` with `logic: MutMoveLogic[T, Static]` which would never unify with `reg`.

What we really need is a rank-2 type, so `MutMoveLogic` is instead `MutMoveLogic[m: Region -> Type] = forall r, (GameState, PlayerId, m[r]) -> Move \ r`. But Flix doesn't have rank-2 types, so we need some kind of workaround.

**Comment by JonathanStarup** [comment]
> I've run into a problem with the Impure to Pure MoveLogic function.
> ```scala
>     type alias MutMoveLogic[m: Region -> Type, r: Region] =
>         (GameState, PlayerId, m[r]) -> Move \ r
> 
>     def toMoveLogic(logic: MutMoveLogic[m, r]): MoveLogic with State[a] = {
>         def f(gs, id, state) = region reg {
>             // call logic() with conversions
>         }
>     }
> ```
> 
> The issue is that here, we want the `r` of `MutMoveLogic[m, r]` to unify with `reg`. But `r` is already bound by the function. For example we could call `toMoveLogic` with `logic: MutMoveLogic[T, Static]` which would never unify with `reg`.
> 
> What we really need is a rank-2 type, so `MutMoveLogic` is instead `MutMoveLogic[m: Region -> Type] = forall r, (GameState, PlayerId, m[r]) -> Move \ r`. But Flix doesn't have rank-2 types, so we need some kind of workaround.

yea I had this problem too. difficult to solve.
